service/src/main/java/bio/terra/workspace/service/workspace/model/GcpCloudContext.java

```diff
@@ -1,61 +1,57 @@
 package bio.terra.workspace.service.workspace.model;
 
 import bio.terra.workspace.common.exception.InternalLogicException;
-import bio.terra.workspace.db.DbSerDes;
 import bio.terra.workspace.db.model.DbCloudContext;
 import bio.terra.workspace.generated.model.ApiGcpContext;
-import bio.terra.workspace.service.resource.model.WsmResourceState;
-import bio.terra.workspace.service.workspace.exceptions.InvalidSerializedVersionException;
 import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
 import com.fasterxml.jackson.annotation.JsonProperty;
-import java.util.Optional;
 import javax.annotation.Nullable;
 
 public class GcpCloudContext implements CloudContext {
-  private final String gcpProjectId;
-  private final String samPolicyOwner;
-  private final String samPolicyWriter;
-  private final String samPolicyReader;
-  private final String samPolicyApplication;
-  private final @Nullable CloudContextCommonFields commonFields;
+  private final @Nullable GcpCloudContextFields contextFields;
+  private final CloudContextCommonFields commonFields;
 
   @JsonCreator
   public GcpCloudContext(
-      @JsonProperty("gcpProjectId") String gcpProjectId,
-      @JsonProperty("samPolicyOwner") String samPolicyOwner,
-      @JsonProperty("samPolicyWriter") String samPolicyWriter,
-      @JsonProperty("samPolicyReader") String samPolicyReader,
-      @JsonProperty("samPolicyApplication") String samPolicyApplication,
-      @JsonProperty("commonFields") @Nullable CloudContextCommonFields commonFields) {
-    this.gcpProjectId = gcpProjectId;
-    this.samPolicyOwner = samPolicyOwner;
-    this.samPolicyWriter = samPolicyWriter;
-    this.samPolicyReader = samPolicyReader;
-    this.samPolicyApplication = samPolicyApplication;
+      @JsonProperty("contextFields") @Nullable GcpCloudContextFields contextFields,
+      @JsonProperty("commonFields") CloudContextCommonFields commonFields) {
+    this.contextFields = contextFields;
     this.commonFields = commonFields;
   }
 
+  @Nullable
+  public GcpCloudContextFields getContextFields() {
+    return contextFields;
+  }
+
+  @JsonIgnore
   public String getGcpProjectId() {
-    return gcpProjectId;
+    return contextFields.getGcpProjectId();
```

Comment:
contextFields is nullable, this needs NPE handling

Reply:
I considered that, but how would I handle it? I thought an NPE would be a good way to know that there was a case where we were doing "get required" and things were not in the right state.

Reply:
I think it would be better to throw a more informative error (like an `InvalidCloudContextStateException`) instead, especially if this gets surfaced to callers. Doing that with other cases in PF-2788 seems reasonable though

Reply:
Checking for null and throwing a `InvalidCloudContextStateException` would be a bit neater, but I'm a lot less concerned about this with most callers using `getRequiredReadyGcpProject` now. At this point hitting that would probably indicate a path where WSM should be calling `getRequiredReady` but isn't